### PR TITLE
Social Notes: Add options and UI for link formatting

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Adds options and UI for link formatting options for Jetpack Social Notes

## Proposed changes:

* Add `social_notes_config` option to Jetpack social options
* Add three options for Jetpack social config

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Go to Jetpack Social admin page
* If Social Notes is disabled, confirm that you do not see the Notes config options 
* Turn ON Social Notes
* Confirm that you see the config options
* Confirm that "Append post link" option is enabled by default.
* Turn OFF "Append post link" option
* Confirm that other two options are hidden
* Change different options
* Reload the page to confirm that the options are persisted.

1. When Social Notes is disabled
    Screenshot 2024-04-01 at 6 04 45 PM

2. When Social Notes is enabled, it shows the default values
    Screenshot 2024-04-01 at 6 04 57 PM

3. When "Append post link" is disabled, it hides the other two options
    Screenshot 2024-04-01 at 6 05 09 PM

4. When all options are enabled
    Screenshot 2024-04-01 at 6 05 23 PM

5. Small screen view
    Screenshot 2024-04-01 at 6 05 33 PM

### CODE REVIEW

**Comment by gmjuhasz** [comment]
I wonder if those descriptions should be below the toggle name, as with the main toggles.

CleanShot 2024-04-02 at 08 55 48 png

I'm not sure what the process for design input with Bebop is, but it would probably be a good idea to ask someone to confirm that the layout is UX-friendly. @pablinos, do you know anything about that?

**Comment by manzoorwanijk** [comment]
> I wonder if those descriptions should be below the toggle name, as with the main toggles.

The current behavior comes from [`@wordpress/components`](https://wordpress.github.io/gutenberg/?path=/story/components-togglecontrol--with-help-text). May be we should use that more often?

**Comment by gmjuhasz** [comment]
> The current behavior comes from [`@wordpress/components`](https://wordpress.github.io/gutenberg/?path=/story/components-togglecontrol--with-help-text). May be we should use that more often?

I think we got prior design work for the admin page, where they came up with the layout we used so far. IMO, it's fine to use the core behavior, but maybe it should be standardized with the main toggles. We can also do that later or seek a design review afterward. The toggles take up too much space with the margins anyway, it was made for just a couple when we started out, it asks for a refactor soon

**Comment by manzoorwanijk** [comment]
> > The current behavior comes from [`@wordpress/components`](https://wordpress.github.io/gutenberg/?path=/story/components-togglecontrol--with-help-text). May be we should use that more often?
> 
> I think we got prior design work for the admin page, where they came up with the layout we used so far. IMO, it's fine to use the core behavior, but maybe it should be standardized with the main toggles. We can also do that later or seek a design review afterward.

Yeah, I agree with having consistent designs.

File: projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js [hunk 1]
```diff
@@ -30,6 +30,28 @@ export function* updateSocialNotesSettings( settings ) {
...
+	const prevConfig = select( SOCIAL_STORE_ID ).getSocialNotesConfig();
+	const newConfig = { ...prevConfig, ...config };
+	try {
+		yield setJetpackSettings( { social_notes_config: newConfig } );
+		yield updateJetpackSettingsControl( { social_notes_config: config } );
+		const updatedSettings = yield fetchJetpackSettings();
+		yield setJetpackSettings( updatedSettings );
+		return true;
+	} catch ( e ) {
+		yield setJetpackSettings( { social_notes_config: newConfig } );
```
Comment (spsiddarthan) [inline projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js]: If this is happening in the catch block, should this be `{ social_notes_config: prevConfig }`?

**Review by spsiddarthan — COMMENTED** [review]
This tests well.

I was able to test the availability of shortlink on sites running just jetpack social as well without the presence of the Jetpack plugin. 

I was able to make use of the `wpme_get_shortlink`  function in fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Sfubegyvax.cuc-og?   D143871-code to get shorturls for my site. 

I think that means if `Use parenthical link format` is enabled, use `Use shortlink` should always be enabled. This is because shortlink citations are recommended to have a short domain name - https://indieweb.org/permashortcitation. 

Alternatively,  `Use parenthical link format` could be a child of the `Use shortlink`.

File: projects/plugins/social/src/js/components/social-notes-toggle/index.tsx [hunk 2]
```diff
@@ -58,6 +69,56 @@ const SocialNotesToggle: React.FC< SocialNotesToggleProps > = ( { disabled } ) =
...
+								checked={ notesConfig.use_shortlink ?? false }
+								className={ styles.toggle }
+								onChange={ onToggleNotesConfig( 'use_shortlink' ) }
+								help={ __(
+									'Whether to use the shortlink instead of the full URL.',
+									'jetpack-social'
+								) }
+							/>
+							<ToggleControl
+								label={ __( 'Use parenthical link format', 'jetpack-social' ) }
```
Comment (spsiddarthan) [inline projects/plugins/social/src/js/components/social-notes-toggle/index.tsx]: Nitpick: Should we simply say `Use permashortcitation`? When I google parenthical link format, I don't get any meaningful results. If I google permashortcitation, the first result is that of the indieweb documentation stub article.

File: projects/plugins/social/src/js/components/social-notes-toggle/index.tsx [hunk 3]
```diff
@@ -58,6 +69,56 @@ const SocialNotesToggle: React.FC< SocialNotesToggleProps > = ( { disabled } ) =
...
+										{ sprintf(
+											/* translators: 1 is the link format */
+											__(
+												'Whether to use the parenthetical link format like %1$s, instead of the standard link.',
+												'jetpack-social'
+											),
+											'(jetpack.com sn/12345)'
+										) }
+										&nbsp;
+										<ExternalLink href="https://jetpack.com/redirect/?source=jetpack-social-notes-help">
```
Comment (spsiddarthan) [inline projects/plugins/social/src/js/components/social-notes-toggle/index.tsx]: We should create a separate task to update the documentation for the shortlink.

**Comment by spsiddarthan** [comment]
> I'm not sure what the process for design input with Bebop is, but it would probably be a good idea to ask someone to confirm that the layout is UX-friendly. @pablinos, do you know anything about that?

We could ping and ask for input from design, but perhaps we roll with what we have and iterate on it? I understand if there are major design changes we will have to work twice on this, perhaps we live with that?

File: projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js (hunk 1 above)
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js]: Oh yes, good catch. Thanks

File: projects/plugins/social/src/js/components/social-notes-toggle/index.tsx (hunk 2 above)
Comment (manzoorwanijk) [reply projects/plugins/social/src/js/components/social-notes-toggle/index.tsx]: Yeah, that's better if it helps users.

File: projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js (hunk 1 above)
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/social-store/actions/social-notes-settings.js]: Fixed. Thanks

File: projects/plugins/social/src/js/components/social-notes-toggle/index.tsx (hunk 2 above)
Comment (manzoorwanijk) [reply projects/plugins/social/src/js/components/social-notes-toggle/index.tsx]: Updated ✅

File: projects/plugins/social/src/js/components/social-notes-toggle/index.tsx (hunk 3 above)
Comment (manzoorwanijk) [reply projects/plugins/social/src/js/components/social-notes-toggle/index.tsx]: Done ✅

**Comment by manzoorwanijk** [comment]
> I was able to make use of the `wpme_get_shortlink` function in fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Sfubegyvax.cuc-og? D143871-code to get shorturls for my site.
> 
> I think that means if `Use parenthical link format` is enabled, use `Use shortlink` should always be enabled. This is because shortlink citations are recommended to have a short domain name - https://indieweb.org/permashortcitation.
> 
> Alternatively, `Use parenthical link format` could be a child of the `Use shortlink`.

May be, instead of createing two boolean toggles, we should create a select/dropdown like “Link format” and let them choose one of the two:
- Full URL
- Shortlink
- Permashortcitation

**Comment by manzoorwanijk** [comment]
Alright, I went ahead with the dropdown after the discussion here p1712120405758179-slack-C02JJ910CNL

Screenshot 2024-04-03 at 1 17 10 PM
Screenshot 2024-04-03 at 1 17 17 PM
Screenshot 2024-04-03 at 1 17 24 PM
Screenshot 2024-04-03 at 1 18 35 PM

**Review by gmjuhasz — APPROVED** [review]
This works well for me, I like the dropdown

**Review by spsiddarthan — APPROVED** [review]

**Pull request merged**